Pillar Tracking

teh60 opened the pull request:
# Key Changes

- JS version 1.2.10&rarr;1.2.11 and Noita Lua Mod version 1.2.8&rarr;1.2.9
- Added Achievement Pillar Tracking (Atreevement Pillars?)
- Added Apotheosis Desert Pillar Tracking

## Credits/Shoutouts

- Thanks to Necauqua for mentioning pillar progress tracking, and for transcribing the pillar image into a YML, though I ended up grabbing the descriptions from noita wiki instead <3
- Thanks to [The Noita Wiki](https://noita.wiki.gg) for having pillar descriptions for both base Noita and Apotheosis' Achievement Pillars

## Noita LUA Mod

- Changed individual mod detection to `ModIsEnabled()` instead of looping through `ModGetActiveModIDs()`
- Added both base Noita and Apotheosis Pillar flag tracking (updates in real time even when the game needs a new run/restart to update the tree in world)

## Onlywands.com Frontend

- Added pillar tracking display
  - Individual pillar quests stay dark until completed
  - Top of the pillar only lights up when whole pillar is complete
  - Overall pillar completion and individual pillar completion displayed above pillars

![image](https://github.com/user-attachments/assets/6c44aa9d-3bdf-4c4c-99e0-3cc5ec32ab98)

- Hovering over any pillar quest zooms the image and gives a tooltip for which pillar it is on, and a (brief) description of the quest

![image](https://github.com/user-attachments/assets/6e2828e5-5fef-4f1a-a769-9bb07c746487)

- Updates as pillar quests are completed in real time

![image](https://github.com/user-attachments/assets/3f5098bc-e7b8-4760-b5dd-9c3c08796a86)

- Also displays Apotheosis' pillars, with a toggle to switch between the two sets of pillars appearing if "Show Apotheosis Content" is enabled

![image](https://github.com/user-attachments/assets/10190c96-d9c3-4ae1-b312-b0ce56d24f44)

## Onlywands.com Backend

- Added pillar progress to `streamerProgress` (previously just perk/spell/enemy progress)
- Added pillars.js and pillarsApoth.js to store pillar icon images, names, and descriptions

Commit:
add pillar progress tracking

Commit:
bump both versions

Commit:
fix logic on pillar brightness on full pillar completion

myndzi commented on public/main.js:
```diff
@@ -1782,6 +1792,124 @@ const containerComp = Vue.component('wands-container', {
+                maxLengths: Math.max(...pillarVersion.map((x) => x.length - 1)),
+            }
+        },
+        progress() {
+            const flags = this.pillarProgress
+            const lengths = Array(this.info.lengths.length).fill(0)
+            const progress = this.info.pillarVersion.map((pillar, i) => pillar.map((flag) => {
+                if (HOP(flag, "key")) {
+                    const inc = flags.includes(flag.key)
+                    lengths[i] += +inc
+                    return inc
+                }
+                return false
+            }))
+            const obj = {
+                // current: Array(this.info.lengths.length).fill(2),
+                current: lengths,
+                total: this.info.lengths,
+            }
+            return {
+                progress,
+                list: obj.total.map((x, i) => {
+                    return {
+                        current: obj.current[i],
+                        total: x,
+                    }
+                }),
+                overall: `${obj.current.reduce((a, c) => a + c, 0)} / ${obj.total.reduce((a, c) => a + c, 0)}`,
+            }
+        },
```
(note: HOP doesn't really need to be used for statically-known keys that are known to be safe. If the property can't be something like __proto__ or toJSON or whatever that exists on the Object prototype, it's fine to just directly access things like flags.key)

The structure of the state vs props vs computed properties here feels weird to me, but I don't think it's worth changing since it already works. Vue 🤮

While I'm often guilty of reaching too frequently for map, reduce, etc. - I find the logic here difficult to follow, especially since map is being used for side effects and with all the nested abstractions. I think breaking it into two functions, clear variable naming, and imperative for loops may help:

```suggestion
        /**
         * Summarize the progress of a single pillar
         * 
         * @param {{key: string, icon: string, name: string}[]} pillarData
         * @return {{progress: boolean[], current: number, total: number}}
         */
        pillarSummary(pillarData) {
          /** @var {string[]} flags */
          const flags = this.pillarProgress;

          let count = 0
          const progress = []
          for (const data of pillarData) {
            const unlocked = flags.includes(data.key)
            progress.push(unlocked)
            if (unlocked) count++
          }

          return {
            progress,
            current: count,
            total: pillarData.length,
          }
        }

        /**
         * Summarize the progress of all pillars
         * 
         * @return {{progress: boolean[][], list: {current: number, total: number}[], overall: string}}
         */
        progress() {
          // the goal here is to return:
          // (1) an array of booleans representing lock/unlock status
          // (2) an array of {current, total} representing the counts of unlocked pillars out of the total
          // (3) a string of the absolute total progress

          /** @var {boolean[][]} progress */
          const allProgress = [];
          /** @var {{current: number, total: number}[]} list */
          const allCounts = [];

          let totalCount = 0;
          let totalTotal = 0;

          for (const pillarData of this.info.pillarVersion) {
            const {progress, ...counts} = this.pillarSummary(pillarData);
            allProgress.push(progress)
            allCounts.push(counts)
            totalCount += counts.current;
            totalTotal += counts.total;
          }

          return {
            progress: allProgress,
            list: allCounts,
            overall: `${totalCount} / ${totalTotal}`,
          };
        }
```

myndzi replied:
Note: TIL about `// @ts-check` - you'll want to add that comment if you want VSCode or whatever to actually tell you about type errors from JSDoc. If you add it to the top of the file it'll do the whole file though, so it might not be super helpful to commit

myndzi reviewed (commented):
The screenshots look really nice!

Left a few questions and comments on some of the code

Commit:
- fixed flags_apoth being unused in utils.lua
- rewrote declaration of inner for-loop in check_flags() for better clarity, filename was me miss-naming and not flag[2] which is the actual filename
- cleaned up pillarsComp info() and progress()

myndzi commented on mod/files/scripts/utils.lua:
```diff
@@ -425,6 +414,50 @@ function get_enemies_progress()
+function get_pillars_progress()
+    local flags =
+    {
+        { { "misc_chest_rain", "crain" }, { "misc_util_rain", "urain" },        { "misc_worm_rain", "wrain" },       { "misc_greed_rain", "grain" },      { "misc_altar_tablet", "train" },        { "misc_mimic_potion_rain", "mrain" },       { "misc_monk_bots", "mbots" },           { "misc_sun_effect", "seffect" }, { "misc_darksun_effect", "dseffect" }, { "secret_tower", "secrett" },       { "playe ...
```
Ah. I got got by 1-indexing LUL

myndzi reviewed (dismissed):
lgtm, thanks for cleaning stuff up :)

Commit:
fixed pillarsApoth.js names being reverse order to keys (noita wiki vs noita apotheosis wiki reversed the order on me)

Commit:
"invert highlighted" now applies to pillar progress

Commit:
fixed "with timer" and "with death trigger" spell tooltips missing information compared to the base spell (i.e. spittter bolt w/ timer was missing information compared to spitter bolt)

Commit:
bumped version

myndzi commented:
LMK when you think it's safe to deploy live.
